### PR TITLE
Qbo wavelet

### PR DESCRIPTION
**Pull request by justin-richling**
## Description

Add additional QBO diagnostics for power spectral density wavelet calculations and plot, see comment from https://github.com/E3SM-Project/e3sm_diags/pull/816#issuecomment-2211114613

- Issue described in #816 

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Review by whannah1 — APPROVED** [review]

File: e3sm_diags/driver/qbo_driver.py [hunk 1]
```diff
@@ -157,6 +158,15 @@ def get_psd_from_deseason(xraw, period_new):
...
 
+def get_psd_from_wavelet(data):
+   deg = 6
+   period = np.arange(1,55+1)
+   freq = 1/period
+   widths = deg / (2*np.pi*freq)
+   cwtmatr = scipy.signal.cwt( data, scipy.signal.morlet2, widths=widths, w=deg )
+   psd = np.mean(np.square(np.abs(cwtmatr)),axis=1)
+   return ( period, psd )
+
```
[the remark this answers is not in the document]
Comment (chengzhuzhang) [reply e3sm_diags/driver/qbo_driver.py:169]: @whannah1 thanks for the review. I agree that inline documentation about the new function will be very helpful!

**Comment by chengzhuzhang** [comment]
Thank you for this PR @justin-richling and @whannah1 
I will review and test soon!

**Comment by chengzhuzhang** [comment]
It looks like I need to update the `json.dump` call to bypass the `TypeError: Object of type int64 is not JSON serializable`: 
-                json.dump(json_dict, outfile)
+                json.dump(json_dict, outfile, default=str)

Also the 4th plot I got seems problematic:
image
@justin-richling could you share the ERA data you are testing with? This looks like a data problem.. Thank you!

**Comment by justin-richling** [comment]
@chengzhuzhang Thanks for pointing this out, the code was getting the power spectra level based on the test case only, not the reference case so I updated this to the `qbo_driver.py` script.

**Comment by justin-richling** [comment]
@chengzhuzhang @tomvothecoder After discussion with @whannah1 it seems for now we will not be including the external wavelet analysis package until a further date. However, I added some clean up and small enhancements to the plot. Here is a snippet of the QBO diags with updated wavet plot:

qbo_diags

With that, this PR should be good to review again, but I seem to be failing the auto checks. I can't seem to narrow down why these small changes are failing but I'm happy to fix them if there easy solutions.

**Comment by chengzhuzhang** [comment]
@justin-richling and @whannah1 thank you for the nice enhancement to identify the value of the peak period. I think we can do something similar to the original method. One thing I would suggest is that the x-axis is also updated and, now the wavelet method has different ticks compare to the original method, as shown below:
Screenshot 2024-09-12 at 2 20 36 PM

Can we consider to revert this part of change to align x-axis for both plots. This makes it easier to compare both methods visually. Let me know what you think...Thanks!

**Comment by whannah1** [comment]
> Can we consider to revert this part of change to align x-axis for both plots. This makes it easier to compare both methods visually.

That seems like a good idea.

**Comment by chengzhuzhang** [comment]
I slightly tweaked the plotting code, so that 3rd and 4th panels have same axis range, and with period of maximum power/variance labeled. The figure is shown below. @justin-richling @whannah1 and @jjbenedict let me know if you'd like any changes. Many thanks for this nice enhancement!

image

**Comment by jjbenedict** [comment]
@chengzhuzhang It's great to see the peak align so closely, thanks for adding this to E3SM Diags!  My only comment:  I think there's a typo in the wavelet y-axis label.  And, variance should be m**2 s**-2...?

**Comment by chengzhuzhang** [comment]
@jjbenedict thank you for catching the error. The units are now fixed with https://github.com/E3SM-Project/e3sm_diags/pull/836/commits/3fd8aa5db8aa35346b33d74f172fe626028b1352

@justin-richling the auto checks are also passed now with some formatting and after merging fixes from `main`. Appreciate your hard work with e3sm_diags!

**Comment by chengzhuzhang** [comment]
@tomvothecoder as a heads-up. We will need to bring this PR to the cdat-migration branch.

**Pull request merged**